libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,573 @@
+*    See Wikipedia for description.
+*
+*    The simplification and pruning algorithms do not alter the path in memory,
+*    and are designed to run in the background. They each have a parameter that
+*    decides how long they will run before saving their state and returning. They
+*    are both anytime algorithms, which is helpful when memory is filling up and
+*    we just need to quickly identify a handful of points which can be deleted.
+*    The algorithms can also report if they are done. If not, the thorough cleanup
+*    will just have to wait for a bit until they are both done.
+*/
+
+AC_SafeRTL::AC_SafeRTL(const AP_AHRS& ahrs) :
+    _ahrs(ahrs),
+    _active(false),
+    _simplification_stack_last_index(-1),
+    _prunable_loops_last_index(-1)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _simplification_bitmask = std::bitset<SAFERTL_PATH_LEN_MAX>().set(); //initialize to 0b1111...
+    _time_of_last_good_position = AP_HAL::millis();
+}
+
+// initialise safe rtl including setting up background processes
+void AC_SafeRTL::init()
+{
+    // check if user has disabled SafeRTL
+    if (_desired_path_len == 0 || is_zero(_accuracy)) {
+        _active = false;
+        return;
+    }
+
+    // protect against repeated call to init
+    if (_initialised) {
+        return;
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _desired_path_len = MIN(SAFERTL_PATH_LEN_MAX, _desired_path_len);
+
+    // allocate arrays
+    _path = (Vector3f*)malloc(_desired_path_len * sizeof(Vector3f));
+    _prunable_loops = (loop*)malloc(_desired_path_len * sizeof(loop));
+    _simplification_stack = (start_finish*)malloc(_desired_path_len * sizeof(start_finish));
+
+    // check if memory allocation failed
+    if (_path == nullptr || _prunable_loops == nullptr || _simplification_stack == nullptr) {
+        log_action(SRTL_DEACTIVATED_INIT_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING, "SafeRTL: unavailable init failed");
+        free(_path);
+        free(_prunable_loops);
+        free(_simplification_stack);
+        _initialised = false;
+        return;
+    } else {
+        _current_path_len = _desired_path_len;
+        _initialised = true;
+    }
+
+    // register SafeRTL cleanup methods to run in IO thread
+    hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AC_SafeRTL::detect_simplifications, void));
+    hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AC_SafeRTL::detect_loops, void));
+}
+
+// clear return path and set home location.  This should be called as part of the arming procedure
+void AC_SafeRTL::reset_path(bool position_ok)
+{
+    Vector3f current_pos;
+    position_ok &= _ahrs.get_relative_position_NED_origin(current_pos);
+    reset_path(position_ok, current_pos);
+}
+
+void AC_SafeRTL::reset_path(bool position_ok, const Vector3f& current_pos)
+{
+    if (!_initialised) {
```

Comment:
I think I've completed an audit of the interface and it seems OK to me.  most functions check "_active" is true and "_active" is only ever true if "_initialised" is also true.  This is because "_reset_path" is the only part of the code that sets "_active" to either _true or _false and this method can only be run if "_initialised" has become true.